### PR TITLE
segment_map: Also segment the bucket array

### PR DESCRIPTION
**Pull request by StephanDollberg**
My change from https://github.com/martinus/unordered_dense/pull/112

Merge into our fork while we sort out the upstreaming part.

### CODE REVIEW

File: include/ankerl/unordered_dense.h [hunk 1]
```diff
@@ -946,7 +949,13 @@ class table : public std::conditional_t<is_map_v<T>, base_table_type_map<T>, bas
         } else {
             m_shifts = other.m_shifts;
             allocate_buckets_from_shift();
-            std::memcpy(m_buckets, other.m_buckets, sizeof(Bucket) * bucket_count());
+            if constexpr (IsSegmented) {
+                for (auto i = 0UL; i < bucket_count(); ++i) {
+                    at(m_buckets, i) = at(other.m_buckets, i);
```
Comment (travisdowns) [inline include/ankerl/unordered_dense.h:954]: Perhaps in the future this could still do the `memcpy` for each segment, right?

**Review by travisdowns — APPROVED** [review]
LGTM

**Review by dotnwat — COMMENTED** [review]
nice. is this the 3rd-party container that was referenced in Travis' issue filed in upstream Seastar regarding a fragmented data structure for use in metrics?

**Comment by StephanDollberg** [comment]
> nice. is this the 3rd-party container that was referenced in Travis' issue filed in upstream Seastar regarding a fragmented data structure for use in metrics?

It would be one of the candidates I guess yes.

File: include/ankerl/unordered_dense.h (hunk 1 above)
Comment (StephanDollberg) [reply include/ankerl/unordered_dense.h:954]: Yes, though we'd need some form of fragment iterator API.

**Pull request merged**